rfcs/convergence/props-typing-improvements.md

```diff
@@ -0,0 +1,148 @@
+# RFC: Improved type checking for component props
+
+Contributors: @behowell
+
+## Summary
+
+This RFC seeks to improve type checking for slot props and props with default values in converged components. It'll make it easier for component authors to "do the right thing" when definining the State type for the component, which reduces the chances for bugs and makes future maintenance easier.
+
+## Problem statement
+
+1. The slot props for a component are mentioned in several different locations in code, without any explicit link between the names and types, other than that they happen to mention the same props.
+2. Some component props are given default values in the `useComponent` hook, but that is not represented in the type system (neither when the default values are provided, nor when they are used in, say `useComponentStyles`).
+
+For example, given a component named `Widget` with slots `icon` and `text`, and a prop `propWithDefaultValue` that is given a default value in `useWidget`:
+
+```typescript
+export interface WidgetProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  icon?: ShorthandProps<IconProps>;
+  text?: ShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;
+
+  // This prop has a default defined in useWidget:
+  propWithDefaultValue?: string;
+
+  // This prop does not have a default value in useWidget (undefined by default)
+  anotherProp?: boolean;
+}
+```
+
+**Problem:** The `WidgeState` type is handwritten, requires repeating the types for slot props and any props that have defaults.
+
+```typescript
+export interface WidgetState extends WidgetProps {
+  icon?: ObjectShorthandProps<IconProps>;
+  text: ObjectShorthandProps<React.HTMLAttributes<HTMLSpanElement>>;

```

Comment:
I also have always questioned this, since we can use the `as` prop... is it really still relevant to declare such an exactly type like `HTMLSpanElement` ?

Reply:
Yeah good point. I agree this should be `React.HTMLAttributes<HTMLElement>`. I wasn't paying too much attention to the details of `Widget` :). That said, we should probably double-check and update existing converged components to see if we should change things like `HTMLSpanElement` to `HTMLElement`